### PR TITLE
Fix typing in `pl.callbacks.lr_monitor`

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7037 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda

### CODE REVIEW

File: pytorch_lightning/utilities/types.py [hunk 1]
```diff
@@ -43,7 +56,30 @@
     Dict[str, Sequence[DataLoader]],
 ]
 EVAL_DATALOADERS = Union[DataLoader, Sequence[DataLoader]]
-# todo: improve LRSchedulerType naming/typing
+
+
+@runtime_checkable
+class _LRScheduler(Protocol):
+    optimizer: Optimizer
+
```
Comment (awaelchli) [inline pytorch_lightning/utilities/types.py:52]: @carmocca 
I couldn't figure out what's the best way to tell mypy that LRScheduler has an attribute "optimizer". Is this the right way?

Comment (carmocca) [reply pytorch_lightning/utilities/types.py:52]: Mmm I don't know for sure.

AFAIK the runtime checkable is for `isinstance` checks
And I don't know if it works with class attributes like that.

Ideally they would define it here:https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.pyi#L4-L10
Or you would extend that

Comment (awaelchli) [reply pytorch_lightning/utilities/types.py]: updated as discussed

Comment (awaelchli) [reply pytorch_lightning/utilities/types.py]: unfortunately, the changes I make here are messing with LightningCLI, not sure yet why

File: pytorch_lightning/utilities/types.py [hunk 2]
```diff
@@ -43,7 +44,75 @@
...
 # todo: improve LRSchedulerType naming/typing
-LRSchedulerTypeTuple = (_LRScheduler, ReduceLROnPlateau)
-LRSchedulerTypeUnion = Union[_LRScheduler, ReduceLROnPlateau]
-LRSchedulerType = Union[Type[_LRScheduler], Type[ReduceLROnPlateau]]
+LRSchedulerTypeTuple = (torch.optim.lr_scheduler._LRScheduler, torch.optim.lr_scheduler.ReduceLROnPlateau)
+LRSchedulerTypeUnion = Union[torch.optim.lr_scheduler._LRScheduler, torch.optim.lr_scheduler.ReduceLROnPlateau]
+LRSchedulerType = Union[Type[torch.optim.lr_scheduler._LRScheduler], Type[torch.optim.lr_scheduler.ReduceLROnPlateau]]
+
+
+class LRSchedulerConfig(TypedDict):
```
Comment (carmocca) [inline pytorch_lightning/utilities/types.py:110]: Since we have this now, we can remove `_get_default_scheduler_config`

File: pytorch_lightning/utilities/types.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/utilities/types.py]: Pushed a commit that fixes the CLI, but breaks mypy, we might need separate variables.

Feel free to revert

File: pytorch_lightning/utilities/types.py [hunk 3]
```diff
@@ -43,7 +44,75 @@
     Dict[str, Sequence[DataLoader]],
 ]
 EVAL_DATALOADERS = Union[DataLoader, Sequence[DataLoader]]
+
+
+# Copied from `torch.optim.lr_scheduler.pyi`
+# Missing attributes were added to improve typing
+class _LRScheduler:
+    optimizer: Optimizer
```
Comment (tchaton) [inline pytorch_lightning/utilities/types.py:52]: Should we open an issue on PyTorch too ?

File: pytorch_lightning/utilities/types.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/types.py:110]: This would be a BC for Flash.

File: pytorch_lightning/utilities/types.py [hunk 4]
```diff
@@ -43,7 +44,75 @@
...
+    def get_lr(self) -> float:
+        ...
+
+    def step(self, epoch: Optional[int] = ...) -> None:
+        ...
+
+
+# Copied from `torch.optim.lr_scheduler.pyi`
+# Missing attributes were added to improve typing
+class ReduceLROnPlateau:
```
Comment (tchaton) [inline pytorch_lightning/utilities/types.py:75]: Is it used ?

**Review by tchaton — APPROVED** [review]
LGTM !  Ideally, this would be done in PyTorch.

File: pytorch_lightning/utilities/types.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/types.py:110]: @carmocca could you elaborate how it is now possible to remove it?

File: pytorch_lightning/utilities/types.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/types.py:75]: yes, below in the LRSchedulerConfig. It is required because in the lr_monitor we access `scheduler_config["scheduler"].optimizer` and mypy needs to know about the optimizer attribute

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/utilities/types.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/utilities/types.py:110]: We cannot set defaults for the `TypedDict`.

Ideally, this would be a dataclass, which then is used in our code. then we don't need the function to create the dict

Comment (awaelchli) [reply pytorch_lightning/utilities/types.py:110]: Right now the code expects a dict everywhere. I think it would be best to leave this for a separate PR in which this can be updated across the code base if needed and provide a way to make it work with flash

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/types.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/types.py:110]: will look into it #11211

File: pytorch_lightning/utilities/types.py [hunk 5]
```diff
@@ -43,7 +44,75 @@
...
+        ...
+
+
 # todo: improve LRSchedulerType naming/typing
-LRSchedulerTypeTuple = (_LRScheduler, ReduceLROnPlateau)
-LRSchedulerTypeUnion = Union[_LRScheduler, ReduceLROnPlateau]
-LRSchedulerType = Union[Type[_LRScheduler], Type[ReduceLROnPlateau]]
+LRSchedulerTypeTuple = (torch.optim.lr_scheduler._LRScheduler, torch.optim.lr_scheduler.ReduceLROnPlateau)
+LRSchedulerTypeUnion = Union[torch.optim.lr_scheduler._LRScheduler, torch.optim.lr_scheduler.ReduceLROnPlateau]
+LRSchedulerType = Union[Type[torch.optim.lr_scheduler._LRScheduler], Type[torch.optim.lr_scheduler.ReduceLROnPlateau]]
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/types.py:107]: @awaelchli quick ques: is there any reason why you didn't use the _LRScheduler and ReduceLROnPlateau types you created here?

Comment (awaelchli) [reply pytorch_lightning/utilities/types.py:107]: something about CLI, I don't know. @carmocca knows

Comment (carmocca) [reply pytorch_lightning/utilities/types.py:107]: Because these are used in the CLI and it needs the actual classes. This is because it uses types to extract the help messages and do input validation
